File: examples/runnable/03-signals-queries-updates/go/approval.go
// Package approval holds the approval lab: signal, query, and update on one
// running execution. The Go equivalent of the Java ApprovalWorkflow lab. The
// Workflow blocks in workflow.Await until an approve/reject Signal arrives, so a
// Query and an Update run against a live execution.
//
// The Workflow definition lives here so both standalone commands can import it:
// ./worker (registers + polls) and ./starter (starts a run).
package approval

import (
	"errors"
	"strings"

	"go.temporal.io/sdk/workflow"
)

const (
	TaskQueue  = "approval"
	WorkflowID = "approval-demo"
)

// ApprovalWorkflow waits for an approve/reject Signal while exposing its state
// via a Query and accepting a validated changeNote Update.
func ApprovalWorkflow(ctx workflow.Context, requestID string) (string, error) {
	status := "WAITING"
	note := "initial request"
	state := func() string {
		return requestID + " " + status + " (" + note + ")"
	}

	if err := workflow.SetQueryHandler(ctx, "currentState", func() (string, error) {
		return state(), nil
	}); err != nil {
		return "", err
	}

	err := workflow.SetUpdateHandlerWithOptions(
		ctx,
		"changeNote",
		func(ctx workflow.Context, newNote string) (string, error) {
			note = newNote
			return state(), nil
		},
		workflow.UpdateHandlerOptions{
			Validator: func(ctx workflow.Context, newNote string) error {
				if strings.TrimSpace(newNote) == "" {
					return errors.New("note is required")
				}
				return nil
			},
		},
	)
	if err != nil {
		return "", err
	}

	approve := workflow.GetSignalChannel(ctx, "approve")
	reject := workflow.GetSignalChannel(ctx, "reject")
	workflow.Go(ctx, func(ctx workflow.Context) {
		selector := workflow.NewSelector(ctx)
		selector.AddReceive(approve, func(c workflow.ReceiveChannel, _ bool) {
			var approver string
			c.Receive(ctx, &approver)
			status = "APPROVED by " + approver
		})
		selector.AddReceive(reject, func(c workflow.ReceiveChannel, _ bool) {
			var reason string
			c.Receive(ctx, &reason)
			status = "REJECTED: " + reason
		})
		selector.Select(ctx)
	})

	_ = workflow.Await(ctx, func() bool {
		return strings.HasPrefix(status, "APPROVED") || strings.HasPrefix(status, "REJECTED")
	})
	return state(), nil
}

File: examples/runnable/03-signals-queries-updates/python/approval.py
"""Approval workflow: signal, query, and update on one running execution.

The Python equivalent of the Java ApprovalWorkflow lab. The Workflow blocks in
workflow.wait_condition until an approve/reject Signal arrives, so a Query and an
Update run against a live execution.
"""

from temporalio import workflow

TASK_QUEUE = "approval"
WORKFLOW_ID = "approval-demo"


@workflow.defn
class ApprovalWorkflow:
    def __init__(self) -> None:
        self._request_id = ""
        self._status = "WAITING"
        self._note = "initial request"

    @workflow.run
    async def run(self, request_id: str) -> str:
        self._request_id = request_id
        await workflow.wait_condition(
            lambda: self._status.startswith("APPROVED")
            or self._status.startswith("REJECTED")
        )
        return self._state()

    @workflow.signal
    def approve(self, approver: str) -> None:
        self._status = "APPROVED by " + approver

    @workflow.signal
    def reject(self, reason: str) -> None:
        self._status = "REJECTED: " + reason

    @workflow.query
    def current_state(self) -> str:
        return self._state()

    @workflow.update
    def change_note(self, note: str) -> str:
        self._note = note
        return self._state()

    @change_note.validator
    def validate_note(self, note: str) -> None:
        if not note or not note.strip():
            raise ValueError("note is required")

    def _state(self) -> str:
        return f"{self._request_id} {self._status} ({self._note})"
